gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,169 @@
+        df.columns = df.columns.set_levels(
+            df.columns.levels[0].str.encode("latin-1").str.decode("utf-8"), level=0
+        )
+        df = df.tz_convert(area.tz)
+        # Truncation will fail if data is not sorted along the index in rare
+        # cases. Ensure the dataframe is sorted:
+        df = df.sort_index(axis=0)
+
+        if df.columns.nlevels == 2:
+            df = (
+                df.assign(newlevel="Actual Aggregated")
+                .set_index("newlevel", append=True)
+                .unstack("newlevel")
+            )
+        df = df.truncate(before=start, after=end)
+        return df
+
+    def _calculate_total_carbon_generated(
+        self, country_code: str, start: pd.Timestamp, end: pd.Timestamp, stat: str = "median"
+    ) -> pd.DataFrame:
+        """Calculate the total carbon generated based on the specified statistic
+        (min, median, max)"""
+        if stat not in ["min", "median", "max"]:
+            raise ValueError("stat must be one of 'min', 'median', or 'max'")
+
+        df = self._query_generation_per_plant(country_code, start, end)
+
+        generation_types = df.iloc[0].iloc[1:]
+        carbon_emission_per_plant = lambda x: GENERATION_PLANT_TO_CARBON_EMISSIONS[x][
+            stat
+        ]  # noqa: E731 C3001
+        emissions_map = generation_types.map(carbon_emission_per_plant)
+
+        df_numeric = df.iloc[2:].reset_index(drop=True)
+        df_numeric = df_numeric.set_index("Unnamed: 0").apply(  # noqa: E501
+            pd.to_numeric, errors="coerce"
+        )
+        df_numeric.index = pd.to_datetime(df_numeric.index)
+        df_numeric["Total Energy (kWh)"] = df_numeric.sum(axis=1) * 1000  # convert MWh to kWh
+        df_numeric["Total Carbon (gCO2eq)"] = df_numeric.mul(emissions_map, axis=1).sum(axis=1)
+
+        generation_types = df.iloc[0]
+        emissions_map = generation_types.map(
+            lambda x: GENERATION_PLANT_TO_CARBON_EMISSIONS.get(x, {}).get(stat, 0)
+        )
+        return df_numeric[["Total Energy (kWh)", "Total Carbon (gCO2eq)"]]
+
+    def update(self, country_code: str, time_slot: duration, results_list: List[Dict]):
+        """Update the results_dict with carbon generated and carbon savings"""
+
+        start_date = pd.Timestamp(results_list[0]["current_market"].split("T")[0])
+        end_date = pd.Timestamp(results_list[-1]["current_market"].split("T")[0])
+        df_total_carbon_and_energy = self._calculate_total_carbon_generated(
+            country_code, start_date, end_date
+        )
+        for result in results_list:
+            current_market = result["current_market"]
+
+            for area_uuid in result["results"].keys():
+                area_result = result["results"][area_uuid]
+                target_timestamp = pd.Timestamp(current_market).tz_localize("UTC")
+
+                nearest_index = df_total_carbon_and_energy.index.get_indexer(
```

Comment:
The term "nearest" confused me at first (i thought you are dealing with locations here). With near you mean: "the closes time stamp to the `target_timestamp`, right?". Maybe you can add an inline comment for clarification.

Reply:
Thanks!